File: CSharp/Kafka/Client/CloudEventHeaders.cs
namespace Kafka.Client;

public class CloudEventHeaders : Dictionary<string, string>
{
    static public void ExtractDefaultHeaders(Dictionary<string,string>? headers,
        out string? specversion, out string? type, out string? source, out string? subject)
    {
        specversion = type = source = subject = null;
        headers?.TryGetValue("specversion", out specversion);
        headers?.TryGetValue("type", out type);
        headers?.TryGetValue("source", out source);
        headers?.TryGetValue("subject", out subject);
    }
    public CloudEventHeaders(string? specVersion = null, string? type = null, string? source = null, string? subject = null)
    {
        Add("specversion", specVersion ?? "");
        Add("type", type ?? "");
        Add("source", source ?? "");
        Add("subject", subject ?? "");
    }

    public string SpecVersion
    {
        get => this["specversion"];
        set => this["specversion"] = value;
    }

    public string Type
    {
        get => this["type"];
        set => this["type"] = value;
    }

    public string Source
    {
        get => this["source"];
        set => this["source"] = value;
    }

    public string Subject
    {
        get => this["subject"];
        set => this["subject"] = value;
    }
}

File: Examples/Python/Kafka/Producer/cloud_event_headers.py

class CloudEventHeaders(dict):
    @staticmethod
    def extract_default_headers(headers: dict=None) -> dict:
        specversion = headers.get('specversion', None) if headers else None
        type = headers.get('type', None) if headers else None
        source = headers.get('source', None) if headers else None
        subject = headers.get('subject', None) if headers else None
        return {
            'specversion': specversion, 
            'type': type,
            'source': source,
            'subject': subject
        }

    def __init__(self, specversion=None, type=None, source=None, subject=None):
        self['specversion'] = specversion
        self['type'] = type
        self['source'] = source
        self['subject'] = subject